storage: use only one mutex for store struct

gyuho opened the pull request:
Follow up from https://github.com/coreos/etcd/pull/4029.

[**`sync.Mutex` is a variable**](https://golang.org/pkg/sync/#Mutex), which means there needs to be only one mutex
value per scope. We don't need a separate mutex inside store struct,
**if we assume that `TxnPut` and `TxnRange` are called ONLY ONCE
per transaction (between `TxnBegin` and `TxnEnd`)**, as [documented](https://github.com/coreos/etcd/blob/master/storage/kv.go#L52-L61):

```go
	// TxnBegin begins a txn. Only Txn prefixed operation can be executed, others will be blocked
	// until txn ends. Only one on-going txn is allowed.
	// TxnBegin returns an int64 txn ID.
	// All txn prefixed operations with same txn ID will be done with the same rev.
	TxnBegin() int64
	// TxnEnd ends the on-going txn with txn ID. If the on-going txn ID is not matched, error is returned.
	TxnEnd(txnID int64) error
	TxnRange(txnID int64, key, end []byte, limit, rangeRev int64) (kvs []storagepb.KeyValue, rev int64, err error)
	TxnPut(txnID int64, key, value []byte) (rev int64, err error)
	TxnDeleteRange(txnID int64, key, end []byte) (n, rev int64, err error)
```

Below are benchmarks before and after (if the delta is negative, it means
performance increase with this patch):

```
BenchmarkStorePut:
[1]:
benchmark               old ns/op     new ns/op     delta
BenchmarkStorePut       51365         55579         +8.20%
BenchmarkStorePut-4     66769         77290         +15.76%

benchmark               old allocs     new allocs     delta
BenchmarkStorePut       26             17             -34.62%
BenchmarkStorePut-4     18             17             -5.56%

benchmark               old bytes     new bytes     delta
BenchmarkStorePut       2000          1523          -23.85%
BenchmarkStorePut-4     1586          1530          -3.53%


[2]:
benchmark               old ns/op     new ns/op     delta
BenchmarkStorePut       53975         40284         -25.37%
BenchmarkStorePut-4     67024         79245         +18.23%

benchmark               old allocs     new allocs     delta
BenchmarkStorePut       26             21             -19.23%
BenchmarkStorePut-4     26             26             +0.00%

benchmark               old bytes     new bytes     delta
BenchmarkStorePut       2001          1725          -13.79%
BenchmarkStorePut-4     2017          2028          +0.55%


[3]:
benchmark               old ns/op     new ns/op     delta
BenchmarkStorePut       49534         33685         -32.00%
BenchmarkStorePut-4     75679         79654         +5.25%

benchmark               old allocs     new allocs     delta
BenchmarkStorePut       20             18             -10.00%
BenchmarkStorePut-4     26             26             +0.00%

benchmark               old bytes     new bytes     delta
BenchmarkStorePut       1666          1571          -5.70%
BenchmarkStorePut-4     2023          2020          -0.15%

---

BenchmarkStoreTxnPut:
[1]:
benchmark                  old ns/op     new ns/op     delta
BenchmarkStoreTxnPut       35277         31067         -11.93%
BenchmarkStoreTxnPut-4     38128         38189         +0.16%

benchmark                  old allocs     new allocs     delta
BenchmarkStoreTxnPut       21             18             -14.29%
BenchmarkStoreTxnPut-4     21             21             +0.00%

benchmark                  old bytes     new bytes     delta
BenchmarkStoreTxnPut       2386          2048          -14.17%
BenchmarkStoreTxnPut-4     2389          2356          -1.38%


[2]:
benchmark                  old ns/op     new ns/op     delta
BenchmarkStoreTxnPut       34149         29346         -14.06%
BenchmarkStoreTxnPut-4     54279         33869         -37.60%

benchmark                  old allocs     new allocs     delta
BenchmarkStoreTxnPut       21             18             -14.29%
BenchmarkStoreTxnPut-4     21             21             +0.00%

benchmark                  old bytes     new bytes     delta
BenchmarkStoreTxnPut       2385          2049          -14.09%
BenchmarkStoreTxnPut-4     2371          2374          +0.13%


[3]:
benchmark                  old ns/op     new ns/op     delta
BenchmarkStoreTxnPut       33314         26506         -20.44%
BenchmarkStoreTxnPut-4     36791         29421         -20.03%

benchmark                  old allocs     new allocs     delta
BenchmarkStoreTxnPut       17             18             +5.88%
BenchmarkStoreTxnPut-4     21             18             -14.29%

benchmark                  old bytes     new bytes     delta
BenchmarkStoreTxnPut       2027          2049          +1.09%
BenchmarkStoreTxnPut-4     2345          2045          -12.79%

```

/cc @xiang90

gyuho commented:
For future reference, test script:

```sh

current_dir=$(pwd)

dot="."
test_function_name="XXX"
# benchmark_function_name="BenchmarkStorePut"
benchmark_function_name="BenchmarkStoreTxnPut"

repeat_size=3

echo "Copying original file..."
cp $HOME/kvstore_old.go $current_dir/kvstore.go

for i in `seq 1 $repeat_size`;
do
	echo "echo 1 > /proc/sys/vm/drop_caches" | sudo sh && cd $HOME/go/pkg && rm -rf *;
	cd $current_dir && fpath="$(echo $current_dir)/old_$i.txt";
	if [ "$benchmark_function_name" == "$dot" ]; then
		echo "running all benchmarks...";
		go test -run $test_function_name -bench . -benchmem -cpu 1,4> $fpath;
	else
		echo "running only $benchmark_function_name";
		go test -run $test_function_name -bench $benchmark_function_name -benchmem -cpu 1,4 > $fpath;
	fi
done

echo "Copying new file..."
cp $HOME/kvstore_new.go $current_dir/kvstore.go

for i in `seq 1 $repeat_size`;
do
	echo "echo 1 > /proc/sys/vm/drop_caches" | sudo sh && cd $HOME/go/pkg && rm -rf *;
	cd $current_dir && fpath="$(echo $current_dir)/new_$i.txt";
	if [ "$benchmark_function_name" == "$dot" ]; then
		echo "running all benchmarks...";
		go test -run $test_function_name -bench . -benchmem -cpu 1,4 > $fpath;
	else
		echo "running only $benchmark_function_name";
		go test -run $test_function_name -bench $benchmark_function_name -benchmem -cpu 1,4 > $fpath;
	fi
done

echo "$(echo $benchmark_function_name):" > $current_dir/benchmark_results.txt;
for i in `seq 1 $repeat_size`;
do
	echo "[$(echo $i)]:" >> $current_dir/benchmark_results.txt;
	old_path="$(echo $current_dir)/old_$i.txt" && new_path="$(echo $current_dir)/new_$i.txt" && \
	benchcmp $old_path $new_path >> $current_dir/benchmark_results.txt && \
	echo "" >> $current_dir/benchmark_results.txt && \
	echo "" >> $current_dir/benchmark_results.txt;
done

echo "Done"
```

gyuho commented:
For the `BenchmakrStorePut`, I think test results vary. Another run:

```
BenchmarkStorePut:
[1]:
benchmark               old ns/op     new ns/op     delta
BenchmarkStorePut       52914         53201         +0.54%
BenchmarkStorePut-4     60608         71636         +18.20%

benchmark               old allocs     new allocs     delta
BenchmarkStorePut       26             18             -30.77%
BenchmarkStorePut-4     26             22             -15.38%

benchmark               old bytes     new bytes     delta
BenchmarkStorePut       2000          1597          -20.15%
BenchmarkStorePut-4     2024          1787          -11.71%


[2]:
benchmark               old ns/op     new ns/op     delta
BenchmarkStorePut       43324         51032         +17.79%
BenchmarkStorePut-4     101002        62680         -37.94%

benchmark               old allocs     new allocs     delta
BenchmarkStorePut       21             26             +23.81%
BenchmarkStorePut-4     23             21             -8.70%

benchmark               old bytes     new bytes     delta
BenchmarkStorePut       1725          2000          +15.94%
BenchmarkStorePut-4     1863          1762          -5.42%


[3]:
benchmark               old ns/op     new ns/op     delta
BenchmarkStorePut       54043         35646         -34.04%
BenchmarkStorePut-4     77141         58240         -24.50%

benchmark               old allocs     new allocs     delta
BenchmarkStorePut       26             21             -19.23%
BenchmarkStorePut-4     26             22             -15.38%

benchmark               old bytes     new bytes     delta
BenchmarkStorePut       1999          1724          -13.76%
BenchmarkStorePut-4     2018          1791          -11.25%

```

All tests were run with:

```
model name	: Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
go version go1.5.2 linux/amd64
linux kernel: 3.19.0-42-generic
```

xiang90 commented:
@gyuho We have two mutex since we want to allow concurrent range query. The current implementation is bad and does not allow concurrent range query anyway. I am fine with removing the additional mutex now. But we also need to keep in mind that we might need to bring it back some day.

xiang90 commented on storage/kvstore_test.go:
```diff
@@ -831,3 +815,64 @@ func (i *fakeIndex) Compact(rev int64) map[revision]struct{} {
+	return rs
+}
+
+func TestTxnPut(t *testing.T) {
```
can you move the actual test above the helper functions?

xiang90 replied:
I am actually not quite sure about what we want to test here. From the implementation, I can see we try to put foo->bar n times and rand key,value n times. What is the goal of this test? Ensure what?

xiang90 commented on storage/kvstore_test.go:
```diff
@@ -831,3 +815,64 @@ func (i *fakeIndex) Compact(rev int64) map[revision]struct{} {
 func (i *fakeIndex) Equal(b index) bool { return false }
+
+func createMultiBytesSlice(size, bytesSize int) [][]byte {
+	p := make(map[string]struct{})
```
why you do not want the duplicate keys?

gyuho replied:
Actually, it doesn't matter. I will simplify the helper function. Thanks,

xiang90 commented on storage/kvstore_bench_test.go:
```diff
@@ -0,0 +1,66 @@
+		keys[i] = make([]byte, 64)
+		rand.Read(keys[i])
+	}
+
```
probably also prepare vals to be consistent with BenchmarkStoreTxnPut?

gyuho replied:
Ok, I will. Thanks,

xiang90 commented on storage/kvstore_bench_test.go:
```diff
@@ -0,0 +1,66 @@
+	}
+}
+
+func BenchmarkStoreTxnPut(b *testing.B) {
```
Can we add a comment to say we have this test to learn about the overhead of Txn begin/end?

gyuho replied:
Sure

xiang90 commented on storage/kvstore_test.go:
```diff
@@ -831,3 +815,64 @@ func (i *fakeIndex) Compact(rev int64) map[revision]struct{} {
+	}
+}
+
+func createMultiBytesSlice(byteSize, sliceSize int) [][]byte {
```
I feel this should be called `createBytesSlice` (Bytes indicates byte array)

xiang90 replied:
bytesLength, sliceLength?

xiang90 commented on storage/kvstore_test.go:
```diff
@@ -831,3 +815,58 @@ func (i *fakeIndex) Compact(rev int64) map[revision]struct{} {
+	s := newStore(tmpPath)
+	defer cleanup(s, tmpPath)
+
+	for i := 0; i < size; i++ {
```
I would prefer n to size. N indicates times, which is more suitable here in my opinion.

gyuho commented:
> But we also need to keep in mind that we might need to bring it back some day.

For sure, I will keep that in mind. Just pushed changes addressing your suggestions. PTAL.

Thanks,

xiang90 commented on storage/kvstore_bench_test.go:
```diff
@@ -0,0 +1,74 @@
+	defer os.Remove(tmpPath)
+
+	// prepare keys
+	keys := make([][]byte, b.N)
```
can we just reuse create bytesSlice here?

gyuho replied:
Sure, just changed.

xiang90 commented:
LGTM

gyuho commented:
Thanks. Merging this.
